Add invariants check to BatchedTensorImpl

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40455 Add batching rule for unsqueeze and transpose
* #40172 Initial vmap frontend API
* **#40171 Add invariants check to BatchedTensorImpl**

It checks that all of the bdims in BatchedTensorImpl are sorted in
order of ascending `level`.

Test Plan:
- Check that nothing breaks in `./build/bin/vmap_test`

Differential Revision: [D22102077](https://our.internmc.facebook.com/intern/diff/D22102077)

zou3519 committed:
Add invariants check to BatchedTensorImpl

It checks that all of the bdims in BatchedTensorImpl are sorted in
order of ascending `level`.

Test Plan:
- Check that nothing breaks in `./build/bin/vmap_test`

[ghstack-poisoned]

zou3519 committed:
Update on "Add invariants check to BatchedTensorImpl"


It checks that all of the bdims in BatchedTensorImpl are sorted in
order of ascending `level`.

Test Plan:
- Check that nothing breaks in `./build/bin/vmap_test`

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/BatchedTensorImpl.cpp:
```diff
@@ -61,6 +62,14 @@ int64_t BatchedTensorImpl::actualDim(int64_t dim, bool wrap_dim) const {
+void BatchedTensorImpl::checkInvariants() const {
+  int64_t prev_level = -1;
+  for (const auto& bdim : bdims_) {
+    TORCH_INTERNAL_ASSERT(bdim.level() > prev_level);
```
Why not have this return true or false and then use assert at the call-site? That'll make it more reusable in other circumstances.

zou3519 replied:
We can always refactor this later to return true or false if there are other call sites that need to handle this in a different way. I don't have any other circumstances in mind for calling this function, do you?

ezyang replied:
It's better as written, because if there are multiple invariants you can give more clear error messages when each one fails.

However, if you want to make this debug only, you're gonna to have some extra ifdef to skip the entire thing.

zou3519 committed:
Update on "Add invariants check to BatchedTensorImpl"


It checks that all of the bdims in BatchedTensorImpl are sorted in
order of ascending `level`.

Test Plan:
- Check that nothing breaks in `./build/bin/vmap_test`

Differential Revision: [D22102077](https://our.internmc.facebook.com/intern/diff/D22102077)

[ghstack-poisoned]
